cache-filter/src/unique_callout.rs

```diff
@@ -129,30 +170,171 @@ pub fn free_callout_lock(
         warn!(
             context_id,
-            "thread ({}): trying to free non-existing callout-lock ({})", root_id, request_key
+            "thread ({}): trying to free non-existing callout-lock ({})", root_id, callout_key
         );
-        return Err(Status::NotFound);
+        return Err(UniqueCalloutError::ProxyFailure(Status::NotFound));
     }
 
-    if let Err(e) = set_shared_data(&request_key, None, None) {
+    if let Err(e) = set_shared_data(&callout_key, None, None) {
         warn!(
             context_id,
             "thread ({}): failed to delete the callout-lock ({}) from shared data: {:?}",
             root_id,
-            request_key,
+            callout_key,
             e
         );
-        return Err(e);
+        return Err(UniqueCalloutError::ProxyFailure(e));
+    }
+    Ok(())
+}
+
+pub fn add_to_callout_waitlist(context: &CacheFilter) -> Result<(), UniqueCalloutError> {
+    let waiters_key = format!("CW_{}", context.cache_key.as_string());
+    let queue_id = resolve_shared_queue(crate::VM_ID, &context.root_id.to_string()).unwrap();
+    if queue_id.is_none() {
+        // without queue id, we can't signal waiting thread to resume processing.
+        return Err(UniqueCalloutError::MQResolveFail(context.root_id));
+    }
+    let callout_waiter = CalloutWaiter {
+        queue_id: queue_id.unwrap(),
+        http_context_id: context.context_id,
+    };
+    info!(
+        context.context_id,
+        "thread({}): trying to add to callout waitlist", context.root_id
+    );
+
+    // Unless there is some internal error, there is a guarantee that 1 thread successfully
+    // writes to the cache. So, keep trying and eventually every thread will win.
+    loop {
+        match get_shared_data(&waiters_key) {
+            Ok((Some(bytes), Some(cas))) => {
+                info!(
+                    context.context_id,
+                    "thread({}): adding to existing waitlist", context.root_id
+                );
+                match bincode::deserialize::<Vec<CalloutWaiter>>(&bytes) {
+                    Ok(mut callout_waiters) => {
+                        callout_waiters.push(callout_waiter.clone());
+                        let serialized_cw =
+                            match bincode::serialize::<Vec<CalloutWaiter>>(&callout_waiters) {
+                                Ok(res) => res,
+                                Err(e) => {
+                                    return Err(UniqueCalloutError::SerializeFail {
+                                        what: "Vec<CalloutWaiter>",
+                                        reason: *e,
+                                    })
+                                }
+                            };
+                        if let Err(Status::CasMismatch) =
+                            set_shared_data(&waiters_key, Some(&serialized_cw), Some(cas))
+                        {
+                            continue;
```

Comment:
It's nice to add a log here so we know `CasMismatch` happened.

Reply:
Added